File: src/main/java/org/xena/cs/bsp/BSPParser.java
/*
package org.xena.cs.bsp;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Paths;

*/
/**
 * Created by Jonathan on 8/20/2016.
 *//*

public class BSPParser {
	
	private final int LUMP_LEAFS = 10;
	private final int LUMP_NODES = 5;
	private final int LUMP_PLANES = 1;
	
	private ByteBuffer stream;
	public BspFile.dheader_t BspHeader;
	
	
	public BspFile.dleaf_t[] Leafs;
	public BspFile.lump_t Lumps = new BspFile.lump_t();
	public BspFile.dnode_t[] Nodes;
	public BspFile.dplane_t[] Planes;
	private byte[] fileBytes;
	
	/// <summary>
	///     It loads the Bsp File into the Buffer.
	/// </summary>
	/// <param name="name">Map Name</param>
	public void LoadFile(String name) {
		try {
			fileBytes = Files.readAllBytes(Paths.get(name));
			stream = ByteBuffer.wrap(fileBytes);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	/// <summary>
	///     It returns the Leaf which is at the specified Point
	/// </summary>
	/// <param name="point">The Point you wan't to get the Leaf for</param>
	/// <returns>Leaf which is at the Point</returns>
	
	public BspFile.dleaf_t GetLeafForPoint(BspFile.Vector point) {
		int node = 0;
		
		BspFile.dnode_t pNode;
		BspFile.dplane_t pPlane;
		
		float d = 0.0f;
		
		while (node >= 0) {
			pNode = Nodes[node];
			pPlane = Planes[pNode.planenum];
			
			d = point.x * pPlane.normal.x + point.y * pPlane.normal.y + point.z * pPlane.normal.z - pPlane.dist;
			
			node = d > 0 ? pNode.children[0] : pNode.children[1];
		}
		
		if ((-node - 1) >= 0 && -node - 1 < Leafs.length) {
			return Leafs[-node - 1];
		} else {
			BspFile.dleaf_t newLeaf = new BspFile.dleaf_t();
			
			newLeaf.area = -1;
			newLeaf.contents = 0;
			return newLeaf;
		}
		
	}
	
	private BspFile.dleaf_t GetLeafForPoint2(BspFile.Vector point) {
		int node = 0;
		
		BspFile.dnode_t pNode;
		BspFile.dplane_t pPlane;
		
		float d = 0.0f;
		
		while (node >= 0) {
			pNode = Nodes[node];
			pPlane = Planes[pNode.planenum];
			
			d = point.x * pPlane.normal.x + point.y * pPlane.normal.y + point.z * pPlane.normal.z - pPlane.dist;
			
			node = d > 0 ? pNode.children[0] : pNode.children[1];
		}
		return (
				(-node - 1) >= 0 && -node - 1 < Leafs.length ?
						Leafs[-node - 1] :
						new BspFile.dleaf_t() {
							area =-1,contents =0
						}
		);
		//return Leafs[-node - 1];
	}
	
	/// <summary>
	///     Checks if something is Visible
	/// </summary>
	/// <param name="vStart">Start Point</param>
	/// <param name="vEnd">End Point</param>
	/// <returns>true if visible false if invisible</returns>
	public boolean IsVisible(BspFile.Vector vStart, BspFile.Vector vEnd) {
		BspFile.Vector start = new BspFile.Vector(vEnd.x, vEnd.y, vEnd.z);
		BspFile.Vector end = new BspFile.Vector(vStart.x, vStart.y, vStart.z);
		
		BspFile.Vector vDirection = end - start;
		BspFile.Vector vPoint = end;
		
		int iStepCount = (int) vDirection.Length();
		
		vDirection /= iStepCount;
		
		BspFile.dleaf_t pLeaf = new BspFile.dleaf_t() {
			area =-1
		};
		
		while (iStepCount > 0) {
			vPoint += vDirection;
			
			pLeaf = GetLeafForPoint(new BspFile.Vector(vPoint.x, vPoint.y, vPoint.z));
			
			if (pLeaf.area != -1) {
				if (
						(pLeaf.contents & 0x1) == 0x1 ||
								(pLeaf.contents & 0x8000000) == 0x8000000) {
					break;
				}
			}
			
			iStepCount--;
		}
		return (pLeaf.contents & 0x1) != 0x1;
	}
	
	public bool Visible(_vec3 vStart, _vec3 vEnd) {
		var vDirection = new _vec3(vEnd.x - vStart.x, vEnd.y - vStart.y, vEnd.z - vStart.z);
		
		
		var iStepCount = (int) Math.Sqrt(vDirection.x * vDirection.x + vDirection.y *
				vDirection.y + vDirection.z * vDirection.z);
		
		vDirection.x /= iStepCount;
		vDirection.y /= iStepCount;
		vDirection.z /= iStepCount;
		
		BspFile.dleaf_t pLeaf;
		
		var i = 0;
		var vPoint = vEnd;
		
		while (i <= iStepCount) {
			vPoint.x -= vDirection.x;
			vPoint.y -= vDirection.y;
			vPoint.z -= vDirection.z;
			
			pLeaf = GetLeafForPoint(vPoint);
			
			if (pLeaf.cluster == 0 && pLeaf.area == 0 && pLeaf.numleafbrushes == 0) {
				if ((pLeaf.contents & 0x1) != 0) {
					return false;
				}
			}
			i++;
		}
		
		return true;
	}
	
	/// <summary>
	///     It Loads BspHeader, Lumps, Leafs, Nodes and Planes. Execute it after every Map Change with LoadFile
	///     Also execute it before using Visible Check
	///     Just needs to be executed if you ran Load File
	/// </summary>
	public void ParseData() {
		byte[] bytes2 = fileBytes;
		using(var br = new BinaryReader(new MemoryStream(bytes2)))
		{
			BspHeader.ident = UtilityReader.ReadInt(stream);
			
			if (BspHeader.ident == (int) ('V' + ('B' << 8) + ('S' << 16) + ('P' << 24)))
				UtilityReader.BigEndian = false;
			else
				UtilityReader.BigEndian = true;
			
			BspHeader.version = UtilityReader.ReadInt(stream);
			
			BspHeader.lumps = new BspFile.lump_t[64];
			for (var i = 0; i < BspHeader.lumps.Length; i++) {
				BspHeader.lumps[i] = new BspFile.lump_t();
				
				BspHeader.lumps[i].fileofs = UtilityReader.ReadInt(stream);
				
				BspHeader.lumps[i].filelen = UtilityReader.ReadInt(stream);
				
				BspHeader.lumps[i].version = UtilityReader.ReadInt(stream);
				
				BspHeader.lumps[i].fourCC = UtilityReader.ReadInt(stream);
			}
			BspHeader.mapRevision = UtilityReader.ReadInt(stream);
			
			Planes = new BspFile.dplane_t[BspHeader.lumps[LUMP_PLANES].filelen / 20];
			Leafs = new BspFile.dleaf_t[BspHeader.lumps[LUMP_LEAFS].filelen / 56];
			Nodes = new BspFile.dnode_t[BspHeader.lumps[LUMP_NODES].filelen / 32];
		}
		
		
		stream.Position = BspHeader.lumps[LUMP_LEAFS].fileofs;
		
		//fileStream.Read(Leafsa, BspHeader.lumps[LUMP_LEAFS].fileofs, BspHeader.lumps[LUMP_LEAFS].filelen);
		
		for (var i = 0; i < BspHeader.lumps[LUMP_LEAFS].filelen / 56; i++) {
			Leafs[i] = new BspFile.dleaf_t();
			
			Leafs[i].contents = UtilityReader.ReadInt(stream);
			
			Leafs[i].cluster = UtilityReader.ReadShort(stream);
			
			Leafs[i].area = UtilityReader.ReadShort(stream);
			
			Leafs[i].flags = UtilityReader.ReadShort(stream);
			
			Leafs[i].mins = new short[3];
			Leafs[i].mins[0] = UtilityReader.ReadShort(stream);
			Leafs[i].mins[1] = UtilityReader.ReadShort(stream);
			Leafs[i].mins[2] = UtilityReader.ReadShort(stream);
			
			
			Leafs[i].maxs = new short[3];
			Leafs[i].maxs[0] = UtilityReader.ReadShort(stream);
			Leafs[i].maxs[1] = UtilityReader.ReadShort(stream);
			Leafs[i].maxs[2] = UtilityReader.ReadShort(stream);
			
			
			Leafs[i].firstleafface = UtilityReader.ReadUShort(stream);
			
			Leafs[i].numleaffaces = UtilityReader.ReadUShort(stream);
			
			Leafs[i].firstleafbrush = UtilityReader.ReadUShort(stream);
			
			Leafs[i].numleafbrushes = UtilityReader.ReadUShort(stream);
			
			Leafs[i].leafWaterDataID = UtilityReader.ReadShort(stream);
		}
		
		
		//var Planesa = fileBytes.Skip(BspHeader.lumps[LUMP_PLANES].fileofs).Take(BspHeader.lumps[LUMP_PLANES].filelen).ToArray();
		//fileStream.Read(Planesa, BspHeader.lumps[LUMP_PLANES].fileofs, BspHeader.lumps[LUMP_PLANES].filelen);
		
		stream.Position = BspHeader.lumps[LUMP_PLANES].fileofs;
		for (var i = 0; i < BspHeader.lumps[LUMP_PLANES].filelen / 20; i++) {
			Planes[i].normal = new BspFile.Vector();
			
			Planes[i].normal.x = UtilityReader.ReadFloat(stream);
			Planes[i].normal.y = UtilityReader.ReadFloat(stream);
			Planes[i].normal.z = UtilityReader.ReadFloat(stream);
			
			Planes[i].dist = UtilityReader.ReadFloat(stream);
			
			Planes[i].type = UtilityReader.ReadInt(stream);
		}
		
		
		//var Nodesa = fileBytes.Skip(BspHeader.lumps[LUMP_NODES].fileofs).Take(BspHeader.lumps[LUMP_NODES].filelen).ToArray();
		//fileStream.Read(Nodesa, BspHeader.lumps[LUMP_NODES].fileofs, BspHeader.lumps[LUMP_NODES].filelen);
		stream.Position = BspHeader.lumps[LUMP_NODES].fileofs;
		
		for (var index = 0; index < BspHeader.lumps[LUMP_NODES].filelen / 32; index++) {
			Nodes[index] = new BspFile.dnode_t();
			
			Nodes[index].planenum = UtilityReader.ReadInt(stream);
			
			Nodes[index].children = new int[2];
			Nodes[index].children[0] = UtilityReader.ReadInt(stream);
			Nodes[index].children[1] = UtilityReader.ReadInt(stream);
			
			
			Nodes[index].mins = new short[3];
			Nodes[index].mins[0] = UtilityReader.ReadShort(stream);
			Nodes[index].mins[1] = UtilityReader.ReadShort(stream);
			Nodes[index].mins[2] = UtilityReader.ReadShort(stream);
			
			
			Nodes[index].maxs = new short[3];
			Nodes[index].maxs[0] = UtilityReader.ReadShort(stream);
			Nodes[index].maxs[1] = UtilityReader.ReadShort(stream);
			Nodes[index].maxs[2] = UtilityReader.ReadShort(stream);
			
			Nodes[index].firstface = UtilityReader.ReadUShort(stream);
			
			Nodes[index].numfaces = UtilityReader.ReadUShort(stream);
			
			Nodes[index].area = UtilityReader.ReadShort(stream);
			
			Nodes[index].paddding = UtilityReader.ReadShort(stream);
		}
	}
	
}
*/

File: src/main/java/org/xena/cs/bsp/BSPParser.kt
package org.xena.cs.bsp

import org.xena.cs.bsp.BspFile.*
import org.xena.plugin.utils.Vector
import org.xena.plugin.utils.ushort
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files
import java.nio.file.Paths


class BSPParser(val fileName: String) {
	
	private val LUMP_LEAFS = 10
	private val LUMP_NODES = 5
	private val LUMP_PLANES = 1
	
	val stream by lazy { ByteBuffer.wrap(Files.readAllBytes(Paths.get(fileName))) }
	
	var header = dheader_t()
	
	var leafs = emptyArray<dleaf_t>()
	var nodes = emptyArray<dnode_t>()
	var planes = emptyArray<dplane_t>()
	
	fun getLeafForPoint(point: Vector): dleaf_t {
		var index = 0
		
		var node: dnode_t
		var plane: dplane_t
		
		while (index >= 0) {
			node = nodes[index]
			plane = planes[node.planenum]
			
			val d = (point.x * plane.normal.x + point.y * plane.normal.y + point.z * plane.normal.z) - plane.dist
			
			index = if (d > 0) node.children[0] else node.children[1]
		}
		if (-index - 1 >= 0 && -index - 1 < leafs.size) {
			return leafs[-index - 1]
		} else {
			val newLeaf = dleaf_t()
			
			newLeaf.area = -1
			newLeaf.contents = 0
			return newLeaf
		}
	}
	
	fun visible(vStart: Vector, vEnd: Vector): Boolean {
		val vDirection = vEnd - vStart
		
		var steps = vDirection.length()
		vDirection /= steps
		
		var leaf: dleaf_t
		val vPoint = vEnd.copy()
		while (steps >= 0) {
			vPoint -= vDirection
			
			leaf = getLeafForPoint(vPoint)
			
			if (leaf.area.toInt() != -1) {
				if (leaf.contents and 0x1 != 0) {
					return false
				}
			}
			steps--
		}
		return true
	}
	
	fun parse() {
		header.ident = stream.int
		
		if (header.ident == 'V'.toInt() + ('B'.toInt() shl 8) + ('S'.toInt() shl 16) + ('P'.toInt() shl 24)) {
			stream.order(ByteOrder.BIG_ENDIAN)
		} else {
			stream.order(ByteOrder.nativeOrder())
		}
		
		header.version = stream.int
		
		header.lumps = Array(64) {
			val lump = BspFile.lump_t()
			lump.fileofs = stream.int
			lump.filelen = stream.int
			lump.version = stream.int
			lump.fourCC = stream.int
			
			lump
		}
		
		header.mapRevision = stream.int
		
		stream.position(header.lumps[LUMP_LEAFS].fileofs)
		leafs = Array(header.lumps[LUMP_LEAFS].filelen / 56) {
			val leaf = dleaf_t()
			
			leaf.contents = stream.int
			leaf.cluster = stream.short
			
			val areaFlags = stream.int
			leaf.area = areaFlags and 0x1FF
			leaf.flags = areaFlags shr 9 and 0x3F
			
			leaf.flags = stream.short.toInt()
			
			for (i in 0..leaf.mins.lastIndex) leaf.mins[i] = stream.short
			for (i in 0..leaf.maxs.lastIndex) leaf.maxs[i] = stream.short
			
			leaf.firstleafface = stream.ushort
			leaf.numleaffaces = stream.ushort
			leaf.firstleafbrush = stream.ushort
			leaf.numleafbrushes = stream.ushort
			leaf.leafWaterDataID = stream.short
			
			leaf
		}
		
		stream.position(header.lumps[LUMP_PLANES].fileofs)
		planes = Array(header.lumps[LUMP_PLANES].filelen / 20) {
			val plane = dplane_t()
			
			plane.normal = Vector(stream.float, stream.float, stream.float)
			plane.dist = stream.float
			plane.type = stream.int
			
			plane
		}
		
		stream.position(header.lumps[LUMP_NODES].fileofs)
		nodes = Array(header.lumps[LUMP_NODES].filelen / 32) {
			val node = dnode_t()
			
			node.planenum = stream.int
			
			for (i in 0..node.children.lastIndex) node.children[i] = stream.int
			for (i in 0..node.mins.lastIndex) node.mins[i] = stream.short
			for (i in 0..node.maxs.lastIndex) node.maxs[i] = stream.short
			
			node.firstface = stream.ushort
			node.numfaces = stream.ushort
			node.area = stream.short
			node.paddding = stream.short
			
			node
		}
	}
	
	companion object {
		
		@JvmStatic fun main(args: Array<String>) {
			val tspawn = Vector(744.60254f, -604.03125f, 220.09381f)
			val tspawn2 = Vector(743.60254f, -604.03125f, 220.09381f)
			val upperTunnels = Vector(-1406.858f, 1254.9688f, 96.09381f)
			val p = BSPParser("E:\\Games\\Steam\\SteamApps\\common\\Counter-Strike Global Offensive\\csgo\\maps\\de_dust2.bsp")
			p.parse()
			
			println(p.visible(tspawn, tspawn2))
			println(p.visible(tspawn, upperTunnels))
			println(p.visible(upperTunnels, tspawn2))
		}
	}
	
}
